### ISSUE TITLE
Stale SCO state for multiple connect_with_options()

### ISSUE THREAD

**Issue by k4vej** [dsuch/pymqi#244]
TL'DR:
If you perform a connect_with_options() more than once, with different SCO() params, all those after the first connection are actually sending the SCO() from the first connection.

I am successfully using PyMQI to connect to 100s of different qmgrs.  There are 3 different certificates used to connect to 3 groups of these qmgrs (i.e. 100 qmgrs use cert1, 100 qmgrs use cert2, 100 qmgrs use cert3)

In a single python process, if I connect to multiple qmgrs that share the same certificate, everything works fine.  However, if I connect to multiple qmgrs which have different certificates, obviously each with their own correctly configured SCO(), the first connect_with_options() works fine, however any subsequent connect_with_options() seems to have some stale references to the SCO() used in the first connect_with_options().

I have checked the SCO being passed in to the connect_with_options() is initialised and referencing the correct kdb, and yet when the connect_with_options is called, I can see from the logs on the qmgr, the certificate being presented is from the SCO which was initialized and used in a previous connection to the previous qmgr.

If I perform exactly the same operations in separate processes everything works fine.  I have spent hours and hours verifying, double checking, reading the docs, reading the PyMQI sources.  My suspicion is something to do with the pymqe.c static PyObject usage here, although its my first time looking at Python C bindings so its a little opaque to me.

Working psuedo-code, 2 qmgrs using same certs:
```python
qmgr1 = pymqi.QueueManager(None)
cd1 = pymqi.CD()
sco1 = pymqi.SCO(KeyRepository=KEY1)

qmgr2 = pymqi.QueueManager(None)
cd2 = pymqi.CD()
sco2 = pymqi.SCO(KeyRepository=KEY1)

qmgr1.connect_with_options('QM1', cd1, sco1)
qmgr2.connect_with_options('QM2', cd2, sco2)
```

Non-working psuedo-code, 2 qmgrs using different certs:
```python
qmgr1 = pymqi.QueueManager(None)
cd1 = pymqi.CD()
sco1 = pymqi.SCO(KeyRepository=KEY1)

qmgr2 = pymqi.QueueManager(None)
cd2 = pymqi.CD()
sco2 = pymqi.SCO(KeyRepository=KEY2)

qmgr1.connect_with_options('QM1', cd1, sco1)
qmgr2.connect_with_options('QM2', cd2, sco2) # <<< PROBLEM, logs on Qmgr show content of sco1 was actually sent
```

Working multi-process psuedo-code example, 2 qmgrs using different certs:
```python
connection_details = [
    [pymqi.QueueManager(None), 'QM1', pymqi.CD(), pymqi.SCO(KeyRepository=KEY1)],
    [pymqi.QueueManager(None), 'QM2', pymqi.CD(), pymqi.SCO(KeyRepository=KEY2)],
]

connection_processes = [Process(target=do_connection, args(*connection_input, ))]
for connection_detail in connection_details]

for p in connection_processes:
    p.start()

for p in connection_processes:
    p.join()
```

You can verify the behaviour, by initialising the SCO() in the parent process, and passing it in as args to the forked child processes, this will experience the same failure mode in the multi-process scenario.

**Comment by dsuch** [comment]
Hi @k4vej, I cannot recall it now but what do MQ docs say about sharing SCO or other objects between subprocesses? Is that expected to work? Thanks.

**Comment by k4vej** [comment]
Hey, thanks for the prompt reply.  Not sure will dig into the docs.

I only raised multi-process as an interesting aside, which was my attempted
avenue of narrowing down the culprit.

Initialising two separate SCO objs in a single process isn't working as
they look like they should.

Will see if there are any docs on IBM side to highlight any limitations of
their C libs with respect to SCO.

**Comment by dsuch** [comment]
Can you add this to connect_with_options, in line 1617?
```python
print('CONNECT', name, self, hex(id(sco)))
```

It will confirm if this is the same SCO object.

**Comment by dsuch** [comment]
I looked over the relevant parts of PyMQI and I cannot see anything that would explicitly cache or otherwise reuse SCO buffers.

At this point I ask you to:

* Upload your full source code for everyone to inspect it
* Confirm in MQ docs that what you do is feasible - it sounds like it should be but maybe there is some fine print somewhere saying that what you witness is to be expected

**Comment by k4vej** [comment]
> Can you add this to connect_with_options, in line 1617?
> 
> ```python
> print('CONNECT', name, self, hex(id(sco)))
> ```
> 
> It will confirm if this is the same SCO object.

It was one of the first things I did, the interpreter shows two different addresses- so the SCO objects are unique.

**Comment by k4vej** [comment]
> * Upload your full source code for everyone to inspect it

It is very simple to replicate (the key database creation is more work!).  I could have given a cleaner example, but thought I would use your own ssl-tls.py example as the basis - basically copy and paste everything in your example and make the duplicate of everything unique, and it shows the failure:

```
# all the put and get removed for some brevity
# python 3 encode() added to all the necessary places for str > bytes
    queue_manager1 = 'QMGR001'
    channel1 = 'TEST.CHANNEL'
    host1 = 'qmgr1.fqdn.net'
    port1 = '1414'
    queue_name1 = 'TEST.QUEUE'.encode('utf-8')
    conn_info1 = '%s(%s)' % (host1, port1)
    ssl_cipher_spec1 = 'TLS_RSA_WITH_AES_256_CBC_SHA'
    key_repo_location1 = 'C:\\src\\pki\\QMGR001.kdb'

    cd1 = pymqi.CD()
    cd1.ChannelName = channel1.encode('utf-8')
    cd1.ConnectionName = conn_info1.encode('utf-8')
    cd1.ChannelType = pymqi.CMQC.MQCHT_CLNTCONN
    cd1.TransportType = pymqi.CMQC.MQXPT_TCP
    cd1.SSLCipherSpec = ssl_cipher_spec.encode('utf-8')

    sco1 = pymqi.SCO()
    sco1.KeyRepository = key_repo_location1.encode('utf-8')

    qmgr1 = pymqi.QueueManager(None)
    print(f"Connecting to {queue_manager1}:{id(cd1)}:{id(sco1)}")
    qmgr1.connect_with_options(queue_manager1, cd1, sco1)
    print(f"Connected to {queue_manager1}")


    queue_manager2 = 'QMGR002'
    channel2 = 'TEST.CHANNEL'
    host2 = 'qmgr2.fqdn.net'
    port2 = '1414'
    queue_name2 = 'TEST.QUEUE'.encode('utf-8')
    conn_info2 = '%s(%s)' % (host2, port2)
    ssl_cipher_spec2 = 'TLS_RSA_WITH_AES_256_CBC_SHA'
    key_repo_location2 = 'C:\\src\\pki\\QMGR002.kdb'

    cd2 = pymqi.CD()
    cd2.ChannelName = channel2.encode('utf-8')
    cd2.ConnectionName = conn_info2.encode('utf-8')
    cd2.ChannelType = pymqi.CMQC.MQCHT_CLNTCONN
    cd2.TransportType = pymqi.CMQC.MQXPT_TCP
    cd2.SSLCipherSpec = ssl_cipher_spec.encode('utf-8')

    sco2 = pymqi.SCO()
    sco2.KeyRepository = key_repo_location2.encode('utf-8')

    qmgr2 = pymqi.QueueManager(None)
    print(f"Connecting to {queue_manager2}:{id(cd2)}:{id(sco2)}")
    qmgr2.connect_with_options(queue_manager2, cd2, sco2)

```

**Comment by k4vej** [comment]
Interestingly, while drafting this, I made a typo in the key_repo_location2 path - and it didn't complain with the expected "MQRC_KEY_REPOSITORY_ERROR" during the connect_with_options()

If you do this with the first SCO() and connect_with_options() it errors out as expected

**Comment by SeyfSV** [comment]
@dsuch,  @k4vej 
I can confirm behaviour described by @k4vej

**Comment by dsuch** [comment]
Alright, and what happens when you disconnect from the first queue manager before connecting to the other one?

**Comment by dsuch** [comment]
Also, please provide the exact PyMQI and MQ version that you use, including the operating systems involved, along with your full Python version.

**Comment by SeyfSV** [comment]
> Alright, and what happens when you disconnect from the first queue manager before connecting to the other one?

Nothing, it works as expected without errors 

Current `master` branch
Python 3.8.5:
```
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
[GCC 9.3.0]
```
Linux:
```
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.1 LTS
Release:	20.04
Codename:	focal
```

**Comment by dsuch** [comment]
Thanks @SeyfSV - can you also add the MQ version?

**Comment by SeyfSV** [comment]
updated^

**Comment by k4vej** [comment]
PyMQI 1.12.0 linked against MQ 9.1.2 client libs
Python 3.6.4
Windows 7 64bit and RHEL 7 PyMQI client
RHEL 7 MQ 9.0 server

**Comment by dsuch** [comment]
@k4vej Please send the MQ detals just like @SeyfSV did it.

**Comment by SeyfSV** [comment]
Client error is:
```
17.12.2020 19:39:44 - Process(1200948.1) User(seyf) Program(python)
                    Host(host) Installation(MQNI09010005)
                    VRMF(9.1.0.5)
                    Time(2020-12-17T16:39:44.365Z)
                    RemoteHost(127.0.0.1(8887))
                    ArithInsert1(414) ArithInsert2(575010)
                    CommentInsert1(TEST)
                    CommentInsert2([Class=]GSKVALMethod::X509[Issuer=]CN=mqtest1[#=]058f9c98bbbdc3cb[Subject=]CN=mqtest1[Class=]GSKVALMethod::PKIX[Issuer=]CN=mqtest1[#=]058f9c98bbbdc3cb[Subject=]CN=mqtest1)
                    CommentInsert3(localhost (127.0.0.1)(8887))
                    
AMQ9633E: Bad SSL certificate for channel 'TEST'.
                    
EXPLANATION:
A certificate encountered during SSL handshaking is regarded as bad for one of
the following reasons:
(a) it was formatted incorrectly and could not be validated
(b) it was formatted correctly but failed validation against the Certification
  Authority (CA) root and other certificates held on the local system
(c) it was found in a Certification Revocation List (CRL) on an LDAP server
(d) a CRL was specified but the CRL could not be found on the LDAP server
(e) an OCSP responder has indicated that it is revoked

The channel is 'TEST'; in some cases its name cannot be determined and so is
shown as '????'. The remote host is 'localhost (127.0.0.1)(8887)'. The channel
did not start.

The details of the certificate which could not be validated are
'[Class=]GSKVALMethod::X509[Issuer=]CN=mqtest1[#=]058f9c98bbbdc3cb[Subject=]CN=mqtest1[Class=]GSKVALMethod::PKIX[Issuer=]CN=mqtest1[#=]058f9c98bbbdc3cb[Subject=]CN=mqtest1'.


The certificate validation error was 575010.
ACTION:
Check which of the possible causes applies on your system. Correct the error,
and restart the channel.

This error might indicate that the remote end of the channel is configured to
send the wrong certificate. Check the certificate label configuration at the
remote end of the channel and ensure that the local key repository contains all
of the necessary CA certificates.

```

**Comment by SeyfSV** [comment]
When I added Queue Manager certificate to the client kdb (second request) I got:
```
17.12.2020 20:02:05 - Process(1206851.1) User(seyf) Program(python)
                    Host(host) Installation(MQNI09010005)
                    VRMF(9.1.0.5)
                    Time(2020-12-17T17:02:05.131Z)
                    RemoteHost(127.0.0.1(8887))
                    CommentInsert1(TEST)
                    CommentInsert2(localhost (127.0.0.1)(8887))

AMQ9642E: No SSL or TLS certificate for channel 'TEST'.

EXPLANATION:
The channel 'TEST' did not supply a certificate to use during SSL or TLS
handshaking, but a certificate is required by the remote queue manager.

The remote host is 'localhost (127.0.0.1)(8887)'.
                    
The channel did not start.
ACTION:
Ensure that the key repository of the local queue manager or MQ client contains
a certificate which is associated with the queue manager or client. If you have
configured a certificate label, check that the certificate exists.

Alternatively, if appropriate, change the remote channel definition so that its
SSLCAUTH attribute is set to OPTIONAL and it has no SSLPEER value set.
```

**Comment by k4vej** [comment]
PyMQI on both Windoze and Linux
```
$ pip show pymqi
Name: pymqi
Version: 1.12.0
```

Windoze PyMQI client:
```
Win 7 Enterprise SP1 64bit

$ python -VV
Python 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)]

$ dspmqver
Version:      9.0.0.4
Level:          p900-004-180504
BuildType:   IKAP - (Production)
Platform:     IBM MQ for Windows (x64 platform)
Mode:         64-bit
O/S:             Windows 7 Enterprise x64 Edition, Build 7601: SP1
InstName:    Installation1
InstDesc:
Primary:       Yes
InstPath:      C:\Program Files\IBM\MQ
DataPath:    C:\ProgramData\IBM\MQ
MaxCmdLevel: 900

```

Linux PyMQI client:
```
$ cat /etc/debian_version
jessie/sid

$ uname -a
Linux dfbef6c6-9c9f-40a8-5067-ca0c 4.4.0-148-generic #174~14.04.1-Ubuntu SMP Thu May 9 08:17:37 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

$ python -VV
Python 3.6.9 (default, Jul 2 2019, 23:32:29)
[GCC 4.8.4]

$ dspmqver
Name: IBM MQ
Version: 9.1.2.0
Level: p912-L190308
BuildType: IKAP - (Production)
Platform: IBM MQ for Linux (x86-64 platform)
Mode: 64-bit
O/S: Linux 4.4.0-148-generic
InstName: MQNI09010200
InstDesc: IBM MQ V9.1.2.0 (Redistributable)
Primary: No
InstPath: /home/app/mqclient/linux.x86_64
DataPath: /home/app
MaxCmdLevel: 912
```

MQ Server
```
$ dspmqver
Name:        IBM MQ
Version:      9.0.0.10
Level:          p900-010-200430
BuildType:  IKAP - (Production)
Platform:    IBM MQ for Linux (x86-64 platform)
Mode:        64-bit
O/S:           Linux 3.10.0-1127.19.1.el7.x86_64
O/S Details: Red Hat Enterprise Linux Server 7.8 (Maipo)
InstName:    Installation1
InstDesc:    
Primary:     Yes
InstPath:    /opt/mqm
DataPath:    /var/mqm
MaxCmdLevel: 900
LicenseType: Production
```

**Comment by k4vej** [comment]
@SeyfSV Yeah I would expect you to get that error.

My setup is slightly different, as the certificates in both my kdb's are issued by a Root CA trusted by all my qmgrs.  So even with the connect_with_options() presenting the incorrect first SCO() to the second connection, it SSL handshakes successfully for me. 
 It is when the qmgr tries to use the CN in the client cert is where it fails for me, as it is presenting CN1 from SCO1 instead of CN2 from SCO2.  But that is just the specifics of how I am using certs in my deployments, the root cause for both of us is same same, just different results.

**Comment by dsuch** [comment]
OK, I have two ideas.

From the perspective of MQI, PyMQI is just a C program whose original source code happens to be in Python. We can thus apply to PyMQI all the information found in MQ docs whenever C programs are listed.

In particular, I am looking at MQCNO now and the "Handle-sharing options", which can be found [here](https://www.ibm.com/support/knowledgecenter/SSFKSJ_9.0.0/com.ibm.mq.ref.dev.doc/q095480_.htm).

Please have a look at connect_with_options again. You did not pass "opts" in which means that the method uses CMQC.MQCNO_NONE by default.

In pyme.c, in pymqe_MQCONNX function, we find this line "cno.Options = (MQLONG)options;". This means that the options previously found in connection_with_options are cast to the handle-sharing options directly as they are.

A few lines earlier in the same C function we can also find the definition of cno. It is a MQCNO_DEFAULT object with its version set to MQCNO_VERSION_5.

Now, what you have is essentially the same process and the same thread as far as C programs go and, in combination with the documentation above, it makes me wonder if perhaps the underlying SCO handler is shared with other connections of the same thread?

We do not know it because MQ is not open source and the documentation is vague. It speaks of "sharing of handles" or "connection and object handles" but it is not clear what handles are involved.

But since it hints at sharing, it would be good for you to try out various cno.Options to check what they can influence.

And here is the other idea. Please add this line right before "rv = pymqe.MQCONNX(..." in method connect_with_options and paste back the output here. I would like to check if self.__handle is the same or not across invocations. Thanks.

```python
print('CONNECT', name, self.__handle, self.__name, hex(id(sco)), hex(id(self.__handle)))
```

**Comment by dsuch** [comment]
Let me confirm it - please add that line to connect_with_options and run it with the original code, with the variant that does not disconnect between calls.

**Comment by SeyfSV** [comment]
```
Connecting to MQTEST:140667760655088:140667761100928
CONNECT b'MQTEST' None None 0x103f1f0 0x90a9f0
Connected to MQTEST
Connecting to MQTEST1:140667753906720:140667753907248
CONNECT b'MQTEST1' None None 0x12221f0 0x90a9f0
```

**Comment by k4vej** [comment]
I got the same results - the __handle is the same address

**Comment by dsuch** [comment]
So there it is, PyMQI uses the double-underscore `__handle` attribute for the connection handle and here it was 0x90a9f0.

This is a class-wide object so same instance of QueueManager object will keep reusing it. We can see in connect_with_options that the same `__handle` is re-used. That could probably explain while disconnecting makes it all go away - it sets `__handle` to None.

How MQ treats exactly such a shared handle we can only speculate but it clearly reuses the same SCO in @k4vej's case.

We have a few ways now, unless you can think of something else.

1. @k4vej can disconnect between the calls if the same QueueManager is to be used
2. @k4vej can use a different QueueManager object for each remote queue manager 
3. We can change `__handle` to `_handle` as I suspect the original author's intention was 20 years ago before I took over this project although there is logic too in saying that each remote queue_manager should be represented by a separate QueueManager instance

The last one is bound to break code that accesses the variable in subclasses.

But I can also see that I created #167 earlier in 2020 so perhaps we could do it in the next year, proclaiming that a year worth of advance warning should be enough? What would you say @SeyfSV ?

**Comment by k4vej** [comment]
> This is a class-wide object so same instance of QueueManager object will keep reusing it. 

Can you elaborate - I am looking at class QueueManager, and __handle is an instance variable?

> 2. @k4vej can use a different QueueManager object for each remote queue manager

Same as above really - already spawning a new QueueManager obj for each connection, each with their own self.__handle instance variable.

> 1. @k4vej can disconnect between the calls if the same QueueManager is to be used
I am trying to do stuff at scale, concurrently, so this would breakdown any parallelism unfortunately, or at the least pin it to being far less inefficient, having to use multi-process each with their own interpreter.

**Comment by dsuch** [comment]
Right, I misspoke when I said that option 2. applies here - `__handle` is a class attribute so no matter how many instances of QueueManager you create, the same `__handle` attribute will be shared by them all.

This in fact means that no matter how many instances you spawn, `__handle` will be shared by them all.

The only way that I can see in this case is to make `__handle` no longer be an attribute shared by all the instances.

However, I can no longer spend more time on this ticket at this time - please either contact me directly on dsuch at zato.io to arrange for commercial support or have a look at the source code yourself / wait for assistance from someone else. Thanks.

**Comment by SeyfSV** [comment]
I think this is MQ error/issue or feature.
Debug and trace show that all parameters passed to MQI MQCONNX function are different.
Also, trace shows that gskit initialized only once with `/tmp/pymqi/client.kdb` and than it receives different Labels for different connections.

@dsuch 3 (change __handle to _handle) It was first that I try after id check - nothing changed

@k4vej can you open PMR and ask about creating multiple connection with different kdb-s?
@k4vej as workaround you can use single kdb file for all connection and provide different Labels for different connections. Some additional info about [Labels](https://www.ibm.com/support/knowledgecenter/en/SSFKSJ_9.0.0/com.ibm.mq.sec.doc/q014340_.htm)


```
 22:03:12.860367  1221121.1           :            Attempting to connect using the client library
 22:03:12.865096  1221121.1      RSESS:000001        0x0010:  71692f63 6c69656e 74000000 00000000  |qi/client.......|
 22:03:12.865096  1221121.1      RSESS:000001        0x0230:  00000000 636c6965 6e740000 00000000  |....client......|
 22:03:12.865110  1221121.1      RSESS:000001        0x0010:  71692f63 6c69656e 74000000 00000000  |qi/client.......|
 22:03:12.865110  1221121.1      RSESS:000001        0x0230:  00000000 636c6965 6e740000 00000000  |....client......|
 22:03:12.881609  1221121.1      RSESS:000001      gsk_attribute_set_buffer: gsk_env_handle=0x1d5fcf0, gsk_buf_id = 201, CharsForBuffer = /tmp/pymqi/client.kdb, Length = 21
 22:03:12.881650  1221121.1      RSESS:000001      gsk_attribute_set_buffer: gsk_env_handle=0x1d5fcf0, gsk_buf_id = 204, CharsForBuffer = /tmp/pymqi/client.sth, Length = 21
 22:03:12.881702  1221121.1      RSESS:000001      MQSCO CertificateLabel 'client'
 22:03:12.881707  1221121.1      RSESS:000001      Saved CertificateLabel 'client'
 22:03:12.881839  1221121.1      RSESS:000001      RetCode = 20009518, rc1 = 1013, rc2 = 0, Comment1 = '/home/seyf/IBM/MQ/data/AMQCLCHL.TAB', Comment2 = '', Comment3= '', File= '/build/slot2/p910_P/src/lib/client/amqrcdfa.c', Line= '618'
 22:03:12.882088  1221121.1      RSESS:000001      Determining signature algorithms for TLS 1.2 clients
 22:03:12.882859  1221121.1      RSESS:000001      SSL/TLS outbound environment CertificateLabel 'client'
 22:03:12.882862  1221121.1      RSESS:000001      gsk_attribute_set_buffer: gsk_env_handle=0x1d5fcf0, gsk_buf_id = 203, CharsForBuffer = client, Length = 6
 22:03:12.960286  1221121.1      RSESS:000001      gsk_attribute_get_cert_by_label: input: gsk_env_handle=0x1d5fcf0, key_label=client, *cert_data =0x1d134e0 *cert_elem_count = 32623
 22:03:12.960322  1221121.1      RSESS:000001      gsk_attribute_get_cert_by_label: output: gsk_env_handle=0x1d5fcf0, key_label=client, *cert_data=0x1dcd570, **cert_data=0x1c, *cert_elem_count=9
 22:03:13.009369  1221121.1      RSESS:000001 ---------{  socket
 22:03:13.009405  1221121.1      RSESS:000001 ---------}! socket rc=Unknown(7) FunctionTime=36
 22:03:13.015797  1221121.1      RSESS:000001      SSL socket handle count: 1
 22:03:13.017583  1221121.1      RSESS:000001      SSL/TLS outbound socket CertificateLabel 'client'
 22:03:13.017602  1221121.1      RSESS:000001      gsk_attribute_set_buffer: gsk_env_handle=0x1e332f0, gsk_buf_id = 203, CharsForBuffer = client, Length = 6
 22:03:13.017793  1221121.1      RSESS:000001      gsk_attribute_get_cert_by_label: input: gsk_env_handle=0x1d5fcf0, key_label=client, *cert_data =(nil) *cert_elem_count = 0
 22:03:13.018012  1221121.1      RSESS:000001      gsk_attribute_get_cert_by_label: output: gsk_env_handle=0x1d5fcf0, key_label=client, *cert_data=0x1dcd570, **cert_data=0x1c, *cert_elem_count=9
 22:03:13.038013  1221121.1      RSESS:000001      Events (1) returned for socket (7)
 22:03:13.039254  1221121.1      RSESS:000001        0x0370:  03550403 1306636c 69656e74 00133011  |.U....client..0.|
 22:03:13.066835  1221121.1      RSESS:000001        0x0040:  1306636c 69656e74 301e170d 32303132  |..client0.. 2012|
 22:03:13.066835  1221121.1      RSESS:000001        0x0070:  55040313 06636c69 656e7430 82012230  |U....client0.."0|
 22:03:13.073400  1221121.1      RSESS:000001      Events (1) returned for socket (7)
 22:03:13.079950  1221121.1      RSESS:000001      Events (1) returned for socket (7)
 22:03:13.082797  1221121.1      RSESS:000001      Current socket receive buffer size: 131072
 22:03:13.082802  1221121.1      RSESS:000001      Current socket send buffer size: 2626560
 22:03:13.082843  1221121.1      RSESS:000001      Final socket receive buffer size: 131072
 22:03:13.082847  1221121.1      RSESS:000001      Final socket send buffer size: 2626560
 22:03:13.084829  1221121.2           :            Events (1) returned for socket (7)
 22:03:13.169135  1221121.2           :            Events (1) returned for socket (7)
 22:03:13.170453  1221121.1           :            Attempting to connect using the client library
 22:03:13.170868  1221121.1      RSESS:000002        0x0010:  71692f63 6c69656e 74310000 00000000  |qi/client1......|
 22:03:13.170868  1221121.1      RSESS:000002        0x0230:  00000000 636c6965 6e743100 00000000  |....client1.....|
 22:03:13.170876  1221121.1      RSESS:000002        0x0010:  71692f63 6c69656e 74310000 00000000  |qi/client1......|
 22:03:13.170876  1221121.1      RSESS:000002        0x0230:  00000000 636c6965 6e743100 00000000  |....client1.....|
 22:03:13.171060  1221121.1      RSESS:000002      MQSCO CertificateLabel 'client1'
 22:03:13.171063  1221121.1      RSESS:000002      Saved CertificateLabel 'client1'
 22:03:13.217320  1221121.1      RSESS:000002 ---------{  socket
 22:03:13.217333  1221121.1      RSESS:000002 ---------}! socket rc=Unknown(8) FunctionTime=13
 22:03:13.219271  1221121.1      RSESS:000002      SSL socket handle count: 2
 22:03:13.219688  1221121.1      RSESS:000002      SSL/TLS outbound socket CertificateLabel 'client1'
 22:03:13.219692  1221121.1      RSESS:000002      gsk_attribute_set_buffer: gsk_env_handle=0x1e4c5e0, gsk_buf_id = 203, CharsForBuffer = client1, Length = 7
 22:03:13.219718  1221121.1      RSESS:000002      gsk_attribute_get_cert_by_label: input: gsk_env_handle=0x1d5fcf0, key_label=client1, *cert_data =(nil) *cert_elem_count = 0
 22:03:13.219790  1221121.1      RSESS:000002      SSL/TLS socket CertificateLabel 'client1' does not exist
 22:03:13.223018  1221121.1      RSESS:000002      Events (1) returned for socket (8)
 22:03:13.223178  1221121.1      RSESS:000002        0x0370:  0e060355 04031307 636c6965 6e743100  |...U....client1.|
 22:03:13.230363  1221121.1      RSESS:000002      Events (1) returned for socket (8)
 22:03:13.236979  1221121.1      RSESS:000002      Events (1) returned for socket (8)
 22:03:13.237629  1221121.1      RSESS:000002      RetCode = 20009642, rc1 = 0, rc2 = 0, Comment1 = '', Comment2 = '', Comment3= '', File= '/build/slot2/p910_P/src/lib/client/amqrcafa.c', Line= '2412'
 22:03:13.237767  1221121.1      RSESS:000002      SSL socket handle count: 1
 22:03:13.245130  1221121.2           :            Events (1) returned for socket (7)
 22:03:13.250074  1221121.2           :            Events (32) returned for socket (7)
 22:03:13.253046  1221121.1           :            SSL socket handle count: 0
```

**Comment by k4vej** [comment]
@SeyfSV That is really useful thanks for the trace - I will indeed raise a PMR and let you know where I get to.  What you have just observed looks like what IBM are describing here:

https://www.ibm.com/support/knowledgecenter/en/SSFKSJ_9.0.0/com.ibm.mq.ref.dev.doc/q095520_.htm

**Comment by SeyfSV** [comment]
I think this is answer for all our questions.
> The TLS information can be specified only on:
>     The first MQCONNX call of the client process, or
>     A subsequent MQCONNX call when all previous TLS connections to the queue manager have been concluded using MQDISC.
> 

@k4vej lets close this issue...

**Comment by k4vej** [comment]
> Right, I misspoke when I said that option 2. applies here - `__handle` is a class attribute so no matter how many instances of QueueManager you create, the same `__handle` attribute will be shared by them all.

I guess I am missing something about the C bindings layer, have not yet fully digested Python C extensions yet.  But in the PyMQI  python class QueueManager, __handle in this case is an instance attribute, not a class attribute, and therefore should not be shared?

> However, I can no longer spend more time on this ticket currently - please either contact me directly on dsuch at zato.io to arrange for commercial support or have a look at the source code yourself / wait for assistance from someone else. Thanks.

Appreciate the time you have already spent looking at this - thank you.

**Comment by k4vej** [comment]
I

> I think this is answer for all our questions.
> 
> > The TLS information can be specified only on:
> > The first MQCONNX call of the client process, or
> > A subsequent MQCONNX call when all previous TLS connections to the queue manager have been concluded using MQDISC.
> 
> @k4vej lets close this issue...

At the very least, some docs / commentary around this issue are worth injecting somewhere; So at least other poor individuals trying to do something similar don't fall victim to hours of head scratching ? 
If need be, I can contribute.

**Comment by SeyfSV** [comment]
I tried code with 9.2.1.0 client - now it does not use Label from second call :( As result you can not use workaround that I propose to you.

**Comment by dsuch** [comment]
@k4vej For sure please open a PMR and request that IBM update documentation that @SeyfSV linked to as today it is not easy to find this information. Missing keywords include "subprocesses", "multi-processing", "multi-threading", "multiple threads". With hundreds of queue managers, I assume that you have enough leverage to make it happen quickly.

As for PyMQI - yes, please feel free to add a relevant comment around this section https://dsuch.github.io/pymqi/examples.html#how-to-use-ssl-tls

Thanks.

**Comment by JoshMcMQ** [comment]
The following APAR is not a solution to the issue, but more of a confirmation that MQ only allowing one active MQSCO per process, and also a warning of possible issues with multi-threaded processes that use the MQSCO structure to control SSL parameters when
connecting to an MQ queue manager.

[IT29565: IBM MQ client app using MQCONNX to make an SSL connection fails with FDC from xehExceptionHandler with rriCopyMQSCO on the stack](https://www.ibm.com/support/pages/apar/IT29565)

**Comment by dsuch** [comment]
WAS, WebLogic etc. all use Java which re-implements everything so they are immune to it. But I just wonder about .NET - does it rely on MQI too?

**Comment by JoshMcMQ** [comment]
Unmanaged .NET uses MQI, Managed does not.

**Comment by dsuch** [comment]
So it is a purely client-side limitation. It seems like a left-over from times when only CICS programs were connecting to MQ, each in its own process, and no one at IBM really cares about it because Java (and now .NET) works fine.

**Comment by k4vej** [comment]
PMR is open with IBM right now - will post any updates here.

**Comment by dsuch** [comment]
Hi @k4vej - has there been perhaps any progress on the PMR that you opened? Thanks.

**Comment by SeyfSV** [comment]
Hi @k4vej , what about PMR?

**Comment by k4vej** [comment]
Sorry for the delay in responding gents, terribly rude of me.

I did indeed get a response from IBM on the topic, copied here for completeness:


> After discussing with our development team we have reached in the following conclusion. 

> One of the design limitations of the current MQ client is that it can only use a single key repository (keystore) per process. The very first connection establishes the configuration: even if subsequent connections call MQCONNX and pass in an MQSCO with a different configuration the original configuration will be used, until the connection is closed.

> Only per-channel SSL/TLS attributes may be changed on a per-connection basis: anything in the MQSCO is fixed by the first connection and is used for all subsequent SSL/TLS connections in that process. There is no facility to specify a key repository per channel. 

> So we do not believe the configuration described will work due to the current MQ design. If it is desired to specify different MQSCOs then it will be necessary to start a separate client process for each MQSCO configuration required. 

So that's that then!
My group have worked around this limitation by using a combined KDB, which contains all client certificates used for the various channels, then specifying the particular certificate label required on the channel definition passed into the connect().

I intend to follow up on this with:
1. A PyMQI docs/examples PR which highlights the necessary workaround
2. An IBM RFE to get the MQ Client functionality improved to cater for a threaded client operating in a highly concurrent performance critical, resource limited environment.

**Comment by dsuch** [comment]
Mass-closing old tickets.

**Comment by JoshMcMQ** [comment]
I noted this is fixed in 9.2.5 of MQ and will be included in upcoming 9.3.  You can tell MQ to let you use multiple key stores from the same process. 

You need to add the following to mqclient.ini SSL stanza:
```
EnvironmentScope=CONNECTION
```

It is documented [here](https://www.ibm.com/docs/en/ibm-mq/9.2?topic=file-ssl-stanza-client-configuration).

Hope this help some that has the same issue.
